### ISSUE TITLE
ClusterData with PPI

### ISSUE THREAD

**Issue by CFF-Dream** [rusty1s/pytorch_geometric#1290]
## ❓ Questions & Help

I processed ppi dataset with ClusterData and CluasterLoader,  and set the parameters according to the paper '' Cluster-GCN: An Efficient Algorithm for Training Deep and Large Graph Convolution Networks '', this paper gave a 99.36% accuracy. But my train、val and test accuracy always be 0.000 like "Epoch: 50, Loss: 1.4540, Train: 0.0000, Val: 0.0000, Test: 0.0000 ", do you know why it happened?Thank you very much!

`import torch
import time
import torch.nn as nn
import torch.nn.functional as F
import os.path as osp
from torch_geometric.datasets import PPI
from ppi_cluster import ClusterData, ClusterLoader
from torch_geometric.nn import SAGEConv, ChebConv
from sklearn.metrics import f1_score

path = osp.join(osp.dirname(osp.realpath(__file__)), '..', 'data', 'PPI')
dataset = PPI(path)
train_dataset = PPI(path, split='train')    #20graphs
val_dataset = PPI(path, split='val')        #2graphs
test_dataset = PPI(path, split='test')      #2graphs

print('Partioning the graph... (this may take a while)')
train_dataset_list = []
val_dataset_list = []
test_dataset_list = []
dataset_list = []
train_dataset_index = test_dataset_index =  val_dataset_index = 0

for data in train_dataset:
    cluster_data = ClusterData(data, 'train', train_dataset_index, num_parts=2, recursive=False,
                           save_dir=dataset.processed_dir)
    loader = ClusterLoader(cluster_data, batch_size=20, shuffle=True,
                       num_workers=0)
    train_dataset_list.append(loader)
    dataset_list.append(loader)
    train_dataset_index += 1
    
for data in test_dataset:
    cluster_data = ClusterData(data, 'test', test_dataset_index, num_parts=2, recursive=False,
                           save_dir=dataset.processed_dir)
    loader = ClusterLoader(cluster_data, batch_size=20, shuffle=True,
                       num_workers=0)
    test_dataset_list.append(loader)
    dataset_list.append(loader)
    test_dataset_index += 1

for data in val_dataset:
    cluster_data = ClusterData(data, 'val', val_dataset_index, num_parts=2, recursive=False,
                           save_dir=dataset.processed_dir)
    loader = ClusterLoader(cluster_data, batch_size=20, shuffle=True,
                       num_workers=0)
    val_dataset_list.append(loader)
    dataset_list.append(loader)
    val_dataset_index += 1

print('Done!')

class Net(torch.nn.Module):
    def __init__(self, in_channels, out_channels):
        super(Net, self).__init__()
        
        dim = 512
        
        self.gcn1 = ChebConv(in_channels, dim, K=1)
        self.lin1 = nn.Linear(in_channels, dim)
        self.bn1 = nn.BatchNorm1d(dim)
        
        self.gcn2 = ChebConv(dim, out_channels, K=1)
        self.lin2 = nn.Linear(dim, out_channels)
        self.bn2 = nn.BatchNorm1d(out_channels)
        
        
    def forward(self, x, edge_index):
        x = F.relu(self.gcn1(x, edge_index) + self.lin1(x))
        x = self.bn1(x)
        x = F.relu(self.gcn2(x, edge_index) + self.lin2(x))
        x = self.bn2(x)
        return F.log_softmax(x, dim=1)

device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
model = Net(dataset.num_features, dataset.num_classes).to(device)
loss_op = torch.nn.BCEWithLogitsLoss()
optimizer = torch.optim.Adam(model.parameters(), lr=0.05)

def train():
    model.train()
    total_loss = 0
    for loader in train_dataset_list:
        for data in loader:
            data = data.to(device)
            optimizer.zero_grad()
            loss = loss_op(model(data.x, data.edge_index), data.y)
            total_loss += float(loss.item())
            loss.backward()
            optimizer.step()
    return total_loss / len(train_dataset_list)

def test(list):
    model.eval()
    ys, preds, acc = [], [] , 0
    for loader in list:
        for data in loader:
            ys.append(data.y)
            with torch.no_grad():
                out = model(data.x.to(device), data.edge_index.to(device))
            preds.append((out>0).float().cpu())
        y, pred = torch.cat(ys, dim=0).numpy(), torch.cat(preds, dim=0).numpy()
        f1 = f1_score(y, pred, average='micro') if pred.sum() > 0 else 0
        acc += f1
    return acc / len(list)

PPI_Cluster_LOSS = []
PPI_Train_ACC = []
PPI_Val_ACC = []
PPI_Test_ACC = []
for epoch in range(1, 100):
    loss = train()
    PPI_Cluster_LOSS.append(loss)
    train_accs = test(train_dataset_list)
    PPI_Train_ACC.append(train_accs)
    val_accs = test(val_dataset_list)
    PPI_Val_ACC.append(val_accs)
    test_accs = test(test_dataset_list)
    PPI_Test_ACC.append(test_accs)
    print('Epoch: {:02d}, Loss: {:.4f}, Train: {:.4f}, Val: {:.4f}, Test: {:.4f}'.format(epoch, loss, train_accs, val_accs, test_accs))

####################
class ClusterData(torch.utils.data.Dataset):
    r"""Clusters/partitions a graph data object into multiple subgraphs, as
    motivated by the `"Cluster-GCN: An Efficient Algorithm for Training Deep
    and Large Graph Convolutional Networks"
    <https://arxiv.org/abs/1905.07953>`_ paper.
    Args:
        data (torch_geometric.data.Data): The graph data object.
        num_parts (int): The number of partitions.
        recursive (bool, optional): If set to :obj:`True`, will use multilevel
            recursive bisection instead of multilevel k-way partitioning.
            (default: :obj:`False`)
        save_dir (string, optional): If set, will save the partitioned data to
            the :obj:`save_dir` directory for faster re-use.
            (default: :obj:`None`)
        log (bool, optional): If set to :obj:`False`, will not log any
            progress. (default: :obj:`True`)
    """
    def __init__(self, data, dataset_name, dataset_index, num_parts, recursive=False, save_dir=None,
                 log=True):
        assert data.edge_index is not None
        #i only modified the code in this place 
        recursive_str = '_recursive' if recursive else ''
        filename = f'partition_ppi_{dataset_name}_[{dataset_index}]_{num_parts}{recursive_str}.pt'

        path = osp.join(save_dir or '', filename)
        if save_dir is not None and osp.exists(path):
            adj, partptr, perm = torch.load(path)
        else:
            if log:  # pragma: no cover
                print('Compute METIS partitioning...')

            (row, col), edge_attr = data.edge_index, data.edge_attr
            adj = SparseTensor(row=row, col=col, value=edge_attr)
            adj, partptr, perm = adj.partition(num_parts, recursive)

            if save_dir is not None:
                torch.save((adj, partptr, perm), path)

            if log:  # pragma: no cover
                print('Done!')

        self.data = self.__permute_data__(data, perm, adj)
        self.partptr = partptr
        self.perm = perm
        self.num_parts = num_parts

    def __permute_data__(self, data, perm, adj):
        data = copy.copy(data)
        num_nodes = data.num_nodes

        for key, item in data:
            if item.size(0) == num_nodes:
                data[key] = item[perm]

        data.edge_index = None
        data.edge_attr = None
        data.adj = adj

        return data

    def __len__(self):
        return self.partptr.numel() - 1

    def __getitem__(self, idx):
        start = int(self.partptr[idx])
        length = int(self.partptr[idx + 1]) - start

        data = copy.copy(self.data)
        num_nodes = data.num_nodes

        for key, item in data:
            if item.size(0) == num_nodes:
                data[key] = item.narrow(0, start, length)

        data.adj = data.adj.narrow(1, start, length)

        row, col, value = data.adj.coo()
        data.adj = None
        data.edge_index = torch.stack([row, col], dim=0)
        data.edge_attr = value

        return data

    def __repr__(self):
        return (f'{self.__class__.__name__}({self.data}, '
                f'num_parts={self.num_parts})')
`
